### PR TITLE
Feature/delete account

### LINKED ISSUE #83: "Delete my account" feature on the app
Description:
As a product owner, I want an option on the app that allows users to delete their account or to request for their data to be deleted in order to comply with legal requirements

Reporter:
James Elgar 

Trello link:
https://trello.com/c/Zq6f09nu/160-delete-my-account-feature-on-the-app

Comments:

### PR DESCRIPTION
**Pull request by piotrstanny**
# Description

Please include a brief summary of what the issue is and what has been done to solve it

Fixes #83 

# Checklist:

## Creator

- [ ] The target branch is dev
- [ ] I have updated the unreleased section of the change log
- [ ] I have reviewed the 'files changed' tab on github to ensure all changes are as expected
- [ ] I have added someone to review the pr

## Reviewer

- [ ] I have verified the above are all completed
- [ ] I have run the code locally to ensure it fufills the requirements of the issue
- [ ] I have reviewed the 'files changed' and commented on any sections which I think are not needed, incorrect or could be improved

## After pull

- [ ] If appropriate I have closed the issue

### CODE REVIEW

File: lib/services/auth.dart [hunk 1]
```diff
@@ -179,11 +177,31 @@ class AuthenticationService {
...
+      if (response.statusCode != 200) {
+        print("Returned false, no error but status not 200");
+        return false;
+      }
+      print("Returned true, account should be deleted");
+      return true;
+    } catch (e) {
+      print("Error");
       return false;
     }
```
Comment (JElgar) [inline lib/services/auth.dart:203]: :+1:

File: lib/viewmodels/account_details_model.dart [hunk 2]
```diff
@@ -6,61 +6,77 @@ import 'package:app/locator.dart';
...
-        name: _name, dob: _dob, orgCode: _orgCode
-    ); 
+        name: _name, dob: _dob, orgCode: _orgCode);
+    setBusy(false);
+    notifyListeners();
+  }
+
+  void delete() async {
+    setBusy(true);
+    await _authenticationService.deleteUserAccount();
```
Comment (JElgar) [inline lib/viewmodels/account_details_model.dart:66]: Can we add some error handling in here?

So for example 
```suggestion
    if (!await _authenticationService.deleteUserAccount()) {
        _dialog_service -> something when wrong message
    } else {
        await _analyticsService.logUserAccountDeleted();
        logout();
    }
```

**Review by JElgar — CHANGES_REQUESTED** [review]
Hello thank you for this @piotrstanny 
This looks great, the service looks perfect but can we add-in a little bit extra error handling to the viewmodel. If you want you could also customize the error depending on the status code in service and then give different messages for different types of errors.

One other git thing is I'm not sure why there are so many file changes to the ios folder?

**Review by JElgar — CHANGES_REQUESTED** [review]
One other thing we should think about is testing. I think it would be nice if we make sure we add tests for each new feature we are working on. Let have a chat about this after Christmas though as I think this is gonna require some thinking?

File: lib/viewmodels/account_details_model.dart (hunk 2 above)
Comment (piotrstanny) [reply lib/viewmodels/account_details_model.dart:66]: Yes definitely agree. And as you say we can explore the topic after Christmas.

**Comment by piotrstanny** [comment]
> One other thing we should think about is testing. I think it would be nice if we make sure we add tests for each new feature we are working on. Let have a chat about this after Christmas though as I think this is gonna require some thinking?

Yes definitely agree. And as you say we can explore the topic after Christmas.

File: test/viewmodel_tests/account_view_model_test.dart [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      when(mockDialogService.showDialog(title: "Error", description: "Sorry there has been an error whilst updating your details.")).thenAnswer((_) => Future.value(true));
+      
+      var model = AccountDetailsViewModel();
+      model.name = "jelgar";
+      model.dob = now;
+      model.save();
+
+      // Verify that updateUserDetails function is called with correct parameters when we save
+      verify(mockAuth.updateUserDetails(name: "jelgar", dob: now));
+      expect(await mockAuth.updateUserDetails(name: "jelgar", dob: now), false);
```
Comment (JElgar) [inline test/viewmodel_tests/account_view_model_test.dart:48]: This expect line is what is actually calling the function. This is going to  need another look

**Pull request merged**